nce/NCE-Pairwise-SM/train.py

```diff
@@ -0,0 +1,374 @@
+    """
+    Caches the first randomly generated word vector for a certain size to make it is reused.
+    """
+    cache = {}
+
+    @classmethod
+    def unk(cls, tensor):
+        size_tup = tuple(tensor.size())
+        if size_tup not in cls.cache:
+            cls.cache[size_tup] = torch.Tensor(tensor.size())
+            cls.cache[size_tup].normal_(0, 0.01)
+        return cls.cache[size_tup]
+
+
+def train_sm():
+    args = get_args()
+    config = args
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = True
+
+    # Set random seed for reproducibility
+    torch.manual_seed(args.seed)
+    if not args.cuda:
+        args.gpu = -1
+    if torch.cuda.is_available() and args.cuda:
+        print("Note: You are using GPU for training")
+        torch.cuda.set_device(args.gpu)
+        torch.cuda.manual_seed(args.seed)
+    if torch.cuda.is_available() and not args.cuda:
+        print("You have Cuda but you're using CPU for training.")
+    np.random.seed(args.seed)
+    random.seed(args.seed)
+
+    dataset_root = os.path.join(os.pardir, 'data', 'TrecQA/')
+    train_iter, dev_iter, test_iter = TRECQA.iters(dataset_root, args.vector_cache, args.wordvec_dir, batch_size=args.batch_size,
+                                                   pt_file=True, device=args.gpu, unk_init=UnknownWordVecCache.unk) #
+
+    index2text = np.array(TRECQA.TEXT_FIELD.vocab.itos)
+
+
+    config.target_class = 2
+    config.questions_num = TRECQA.VOCAB_NUM
+    config.answers_num = TRECQA.VOCAB_NUM
+
+    print("index2text:", index2text)
+    print("Dataset {}    Mode {}".format(args.dataset, args.mode))
+    print("VOCAB num", TRECQA.VOCAB_NUM)
+    print("LABEL.target_class:", config.target_class)
+
+    if args.resume_snapshot:
+        if args.cuda:
+            pw_model = torch.load(args.resume_snapshot, map_location=lambda storage, location: storage.cuda(args.gpu))
+        else:
+            pw_model = torch.load(args.resume_snapshot, map_location=lambda storage, location: storage)
+    else:
+        model = SmPlusPlus(config)
+        model.static_question_embed.weight.data.copy_(TRECQA.TEXT_FIELD.vocab.vectors)
+        model.nonstatic_question_embed.weight.data.copy_(TRECQA.TEXT_FIELD.vocab.vectors)
+        model.static_answer_embed.weight.data.copy_(TRECQA.TEXT_FIELD.vocab.vectors)
+        model.nonstatic_answer_embed.weight.data.copy_(TRECQA.TEXT_FIELD.vocab.vectors)
+
+        if args.cuda:
+            model.cuda()
+            print("Shift model to GPU")
+
+        pw_model = PairwiseConv(model)
+
+    parameter = filter(lambda p: p.requires_grad, pw_model.parameters())
+
+    # the SM model originally follows SGD but Adadelta is used here
+    optimizer = torch.optim.Adadelta(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=args.eps)
+    # A good lr is required to use in the following optimizer
+    # optimizer = torch.optim.Adam(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-8)
```

Comment:
Should we make them configurable through command line args instead of commenting them out?

Reply:
Good suggestion! thx.